addons/payment_adyen/controllers/main.py

```diff
@@ -1,37 +1,359 @@
+import hashlib
+import hmac
 import json
 import logging
 import pprint
+
 import werkzeug
+from werkzeug import urls
 
-from odoo import http
+from odoo import _, http
+from odoo.exceptions import ValidationError
 from odoo.http import request
+from odoo.tools.pycompat import to_text
+
+from odoo.addons.payment import utils as payment_utils
+from odoo.addons.payment_adyen.const import CURRENCY_DECIMALS
 
 _logger = logging.getLogger(__name__)
 
 
 class AdyenController(http.Controller):
-    _return_url = '/payment/adyen/return/'
-
-    @http.route([
-        '/payment/adyen/return',
-    ], type='http', auth='public', csrf=False)
-    def adyen_return(self, **post):
-        _logger.info('Beginning Adyen form_feedback with post data %s', pprint.pformat(post))  # debug
-        if post.get('authResult') not in ['CANCELLED']:
-            request.env['payment.transaction'].sudo().form_feedback(post, 'adyen')
-        return werkzeug.utils.redirect('/payment/process')
-
-    @http.route([
-        '/payment/adyen/notification',
-    ], type='http', auth='public', methods=['POST'], csrf=False)
-    def adyen_notification(self, **post):
-        tx = post.get('merchantReference') and request.env['payment.transaction'].sudo().search([('reference', 'in', [post.get('merchantReference')])], limit=1)
-        if post.get('eventCode') in ['AUTHORISATION'] and tx:
-            states = (post.get('merchantReference'), post.get('success'), tx.state)
-            if (post.get('success') == 'true' and tx.state == 'done') or (post.get('success') == 'false' and tx.state in ['cancel', 'error']):
-                _logger.info('Notification from Adyen for the reference %s: received %s, state is %s', states)
+
+    @http.route('/payment/adyen/acquirer_state', type='json', auth='public')
+    def adyen_acquirer_state(self, acquirer_id):
+        """ Return the state of the acquirer.
+
+        :param int acquirer_id: The acquirer handling the transaction, as a `payment.acquirer` id
+        :return: The state of the acquirer: 'enabled' or 'test'
+        :rtype: str
+        """
+        acquirer_sudo = request.env['payment.acquirer'].sudo().browse(acquirer_id).exists()
+        return acquirer_sudo.state
+
+    @http.route('/payment/adyen/payment_methods', type='json', auth='public')
+    def adyen_payment_methods(self, acquirer_id, amount=None, currency_id=None, partner_id=None):
+        """ Query the available payment methods based on the transaction context.
+
+        :param int acquirer_id: The acquirer handling the transaction, as a `payment.acquirer` id
+        :param float amount: The transaction amount
+        :param int currency_id: The transaction currency, as a `res.currency` id
+        :param int partner_id: The partner making the transaction, as a `res.partner` id
+        :return: The JSON-formatted content of the response
+        :rtype: dict
+        """
+        acquirer_sudo = request.env['payment.acquirer'].sudo().browse(acquirer_id)
```

Comment:
is this queriable by anyone on purpose?

Reply:
Payments can be made by the public user so we need to keep the auth as it is.
Now, we could maybe pass an access token to this route to make sure that only verified users can query Adyen's payment methods but, again, I don't think that it's sensitive data. This is just a dict of payment method scheme descriptions. E.g., `{'visa': {'reccurring': True, 'card_length': 16, ...}, ...}`